src/hybrid_crypto/mod.rs

```diff
@@ -1,7 +1,104 @@
+#![allow(non_snake_case)]
+
+use crate::{
+    asymmetric::{ristretto::X25519Crypto, AsymmetricCrypto, KeyPair},
+    symmetric_crypto::{aes_256_gcm_pure::Aes256GcmCrypto, SymmetricCrypto},
+    Error,
+};
+use std::sync::Mutex;
+
 mod block;
+mod dem;
 mod header;
+mod kem;
 mod scanner;
 
 pub use block::Block;
 pub use header::{Header, Metadata};
+use rand_core::{CryptoRng, RngCore};
 pub use scanner::BytesScanner;
+
+/// Key Encapsulation Method (KEM). It is used to generate a secret key along
+/// with its encapsulation. This secret key is then typically used in a DEM
+/// scheme.
+///
+/// TODO: should the KDF used be specified here?
+pub trait Kem: AsymmetricCrypto {
+    /// Length of the sescret key
+    const KEY_LENGTH: usize;
+
+    /// Describe the scheme in plaintext
+    fn description() -> String;
+
+    /// Generate an asymmetric key pair
+    fn key_gen<R: RngCore + CryptoRng>(rng: &Mutex<R>) -> Self::KeyPair;
+
+    /// Generate the secret key and its encapsulation.
+    ///
+    /// - `pk`  : public key
+    fn encaps<R: RngCore + CryptoRng>(
+        rng: &Mutex<R>,
+        pk: &<<Self as AsymmetricCrypto>::KeyPair as KeyPair>::PublicKey,
+    ) -> Result<(Vec<u8>, Vec<u8>), Error>;
+
+    /// Generate the keying data from the given ciphertext and private key.
```

Comment:
```suggestion
    /// Generate the keying data from the given encapsulation and private key.
```
or maybe we don't care because it's the same?